### DISCUSSION TITLE
macOS advisory in case of build failure

### DISCUSSION THREAD

**Discussion by timmyjose** [microsoft/verona#501] [Ideas]
(Putting this here since there is no Wiki for this project)

In case you see an error during build like so:

```
ninja: error: '/Applications/Xcode_11.3.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/lib/libcurses.tbd', needed by 'bin/verona-mlir', missing and no known rule to make it
```

This is due to the fact that `Xcode` is installed as `Xcode.app`, but the build process appends the specific version to the path for some reason. The simplest (hacky) way to bypass this issue to to search and replace `Xcode_11.3.1` (or whichever specific version you have) with `Xcode` inside the build directory.

Note: A better solution would be to inspect the build process and see where this version injection is taking place, but maybe someone else can take that up if it becomes a regular problem.

For reference, my local system is macOS Mojave 10.4.

**Comment by davidchisnall** [comment]
Thanks.  I've seen this before.  It is caused by the current hacky code that tries to download the CI builds of LLVM and use them to avoid needing to rebuild LLVM.  Unfortunately, these embed both OS-version and XCode-version specific paths and so don't actually work on any Mac other than the CI machines (which are set up in their own very special way).  The vcpkg-wip branch should work better for these scenarios eventually: it uses vcpkg to download things and so will be able to get cached builds of LLVM but only if the compiler and other dependency versions exactly match between the cache and the current build system.

**Comment by timmyjose** [comment]
@davidchisnall  That makes a lot of sense. :-) Yes, I did realise that since the project is in constant churn, these things would eventually get sorted out in any case. Just thought of putting it in here in case some others wish to try building it locally and they run into this issue. Funnily enough, I did try checking out the Azure pipelines CI in case the automated builds generated some artifacts, but I couldn't find any, but building it wasn't too much of a hassle at all given the nascent state of the project.

**Comment by rengolin** [comment]
Alternatively, if we can get different MacOS versions in Azure, we could add more LLVM build types and upload the current versions, too.

**Reply by timmyjose** [reply]
That sounds like a great idea for the interim. 👍

**Reply by davidchisnall** [reply]
I don't think it will work.  The Azure Pipelines images have some quite special XCode install configurations that seem to place things in different places to the same version of macOS installed on my laptop.
